### DISCUSSION TITLE
What is the `variables` field in the `ThemeDefinition` object for?

### DISCUSSION THREAD

**Discussion by frederikheld** [vuetifyjs/vuetify#18883] [Q&A] [answered]
[According to the docs](https://vuetifyjs.com/en/features/theme), the `ThemeDefinition` object has an optional field `variables`. It only allows `string | number | undefined`, so it doesn't look like I can use it to define whatever variable I'd like to. What its intended use case?

Can I use it to set variables in my theme that are not colors? Example: in some themes, I'd like to use a specific css effect that is implemented in my Vue app, in others I don't want to use it. The effect works with every theme, but for some themes it just does not fit the style. It would be great if I could turn it off/on via the `ThemeDefinition` like this: `{ variables.useEffect: false }`.

**Comment by KaelWD** [comment, marked as answer]
It sets [https://developer.mozilla.org/en-US/docs/Web/CSS/--*](https://developer.mozilla.org/en-US/docs/Web/CSS/--*)

**Reply by frederikheld** [reply]
Thanks a lot for the clarification. This might even work for my use case.

This info is missing in the docs though.

**Reply by Bejasc** [reply]
@frederikheld how did you go with this? My use-case is very similar to yours, I would like to set an image along side (or preferably directly part of) my theme - but I'm having trouble reading the var

```ts
	theme: {
		defaultTheme: "primary",
		themes: {
			primary: {
				dark: true,

				colors: {
					primary: "#FFD100",
				},
				variables: {
					heroImg: "url(./assets/img/wallpaper2.jpg)",
				},
			},
		},
	},
```
```css
.hero {
	background-image: var("heroImg");
	background-size: cover;
	height: 100vh;
	width: 100vw;
}
```

**Reply by frederikheld** [reply]
Hey @Bejasc, I didn't go further with this, but looking into your code, I see an issue: as the CSS docs linked by Kael state, css vars always start with two dashes. So you should try to access your var with `--heroImg`. In vuetify, theme vars usually start with `--v-theme-`, so you might actually need to use `--v-theme-heroImg`. Just try both and see which one works.

I'm also not quite sure if you can save the whole `url( url_string )` statement in a CSS variable. Maybe you can only store the `url_string` part there. You have to consult the CSS docs about this.

**Reply by Bejasc** [reply]
Thanks @frederikheld, I'd tried many variations including `--`as well as different positions for the `url()` and `var()` without much success so far. 

I'll try another couple of things this morning, but I'm not sure if the variables as part of the theme configuration are actually exposed in this way.

**Reply by KaelWD** [reply]
`var(--v-heroImg)`

**Reply by Bejasc** [reply]
Thanks @KaelWD - I was glad to figure this out earlier.

In case anyone stumbles across this, I'll leave it here, as this thread is one of the few hits `ThemeDefinition` `variables`, as it's largely missing from the docs. 

```ts
const vuetify = createVuetify({
	components,
	directives,
	blueprint: md2,
	theme: {
		defaultTheme: "primary",
		themes: {
			primary: {
				dark: true,
				colors: {
					primary: "#5B729D",
				},
				variables: {
					heroImage: "url('src/assets/img/wallpaper.png')",
					windowColor: "rgba(0,0,30,0.5)",
					windowBackdrop: "blur(6px)",
				},
			},
		},
	},
});
```

```html
<style>
.bg {
	background-color: var(--v-windowColor);
	backdrop-filter: var(--v-windowBackdrop);
}

.hero {
	background-image: var(--v-heroImage);
	background-size: cover;
	height: 100vh;
	width: 100vw;
}
</style>
```

---

The only thing I'm slightly uncertain of in this approach (Which I've now started to use much more widely) is whether or not the image path will hold up in a built/production state, given it needs to reference `src` specifically.

I haven't tested that yet. Using the absolute URL to any public image does work, however. 

I noticed that providing the image as `url('./assets/img/wallpaper.png')` (see `./assets`) in the variable did not resolve the path to the image correctly, whereas using this directly in the style _would_ find the image.

**Reply by Bejasc** [reply]
Referencing https://vitejs.dev/guide/assets (Thanks to Kael for providing via Discord), see [#the-public-directory](https://vitejs.dev/guide/assets#the-public-directory)

moving the image from `src/assets/img/` to `public/img/` and referencing as `"url('image/wallpaper-1.jpg')",` - I'm comfortable that the pathing will remain intact across builds.

**Reply by jacekkarczmarczyk** [reply]
you'd rather import the image instead of putting it in /public
see https://skirtles-code.github.io/vue-examples/guides/working-with-image-assets
